### PR TITLE
DB-10368 Add SPLIT_PART function

### PR DESCRIPTION
**Pull request by ipraznik-splice**

### CODE REVIEW

**Comment by ipraznik-splice** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by dgomezferro — APPROVED** [review]

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/sqlgrammar.jj [hunk 1]
```diff
@@ -18002,6 +18038,7 @@ reservedKeyword() :
 |    tok = <XMLSERIALIZE>
 |    tok = <XMLEXISTS>
 |    tok = <XMLQUERY>
+|    tok = <SPLIT_PART>
```
Comment (arnaud-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/sqlgrammar.jj:18041]: Should that be a reserved keyword? I would rather have added it in the non reserved keyword list.

File: db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java [hunk 2]
```diff
@@ -2644,6 +2645,57 @@ public ConcatableDataValue substring(
...
+        if (this.isNull() ||
+                delimiter == null || delimiter.isNull() || delimiter.getString() == null ||
+                fieldNumber == null || fieldNumber.isNull())
+        {
+            stringResult.setToNull();
+            return stringResult;
+        }
+
+        String valueParts[] = getString().split(Pattern.quote(delimiter.getString()));
+        int partIndex = fieldNumber.getInt();
```
Comment (hatyo) [inline db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java:2686]: what happens if `partIndex` is negative?

**Review by arnaud-splice — CHANGES_REQUESTED** [review]

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/TernaryOperationIT.java [hunk 3]
```diff
@@ -484,5 +481,49 @@ public void testSTRIPNegative() throws Exception {
...
         }
     }
+
+    private void executeAndCheckSplitPart(int fieldNumber) throws SQLException {
+        int count = 0;
+        String sCell1 = null;
+        String sCell2 = null;
+        ResultSet rs;
+
+        rs = methodWatcher.executeQuery(String.format("select split_part(e, ' ', %d), e from ", fieldNumber) + tableWatcherB);
```
Comment (hatyo) [inline splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/TernaryOperationIT.java:491]: Could you please rewrite:
`String.format("select split_part(e, ' ', %d), e from ", fieldNumber) + tableWatcherB`
to:
`String.format("select split_part(e, ' ', %d), e from %s", fieldNumber, tableWatcherB)`

**Review by hatyo — CHANGES_REQUESTED** [review]

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/sqlgrammar.jj (hunk 1 above)
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/sqlgrammar.jj]: done

File: db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java (hunk 2 above)
Comment (ipraznik-splice) [reply db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java:2686]: done

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/TernaryOperationIT.java (hunk 3 above)
Comment (ipraznik-splice) [reply splice_machine/src/test/java/com/splicemachine/derby/impl/sql/execute/operations/TernaryOperationIT.java]: done

**Review by arnaud-splice — APPROVED** [review]

**Review by hatyo — APPROVED** [review]
Good job!

**Review by carolp-503 — APPROVED** [review]
+1

**Pull request merged**